machinery-helper/rpm/rpm.go

```diff
@@ -0,0 +1,162 @@
+	"os"
+	"os/exec"
+	"strings"
+	"unicode/utf8"
+)
+
+// Parses an output line from the rpm command.
+func ParseRpmLine(line string) (fileType string, fileName string, linkTarget string) {
+	fileType = line[0:1]
+
+	index := strings.Index(line, "/")
+	if index < 0 {
+		panic(line)
+	}
+	file := line[index:]
+	fields := strings.Split(file, " -> ")
+	if len(fields) == 2 {
+		fileName = fields[0]
+		linkTarget = fields[1]
+	} else {
+		fileName = file
+	}
+	return
+}
+
+// Returns a list of rpms handled by the system.
+func GetRpms() []string {
+	cmd := exec.Command("rpm", "-qlav")
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := cmd.Run()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	f := func(c rune) bool {
+		return c == '\n'
+	}
+	packages := strings.FieldsFunc(out.String(), f)
+	return packages
+}
+
+// Appends implicit managed dirs to a map.
+func AddImplicitlyManagedDirs(dirs map[string]bool, files map[string]string) {
+	for file, target := range files {
+		for i := 1; i < len(file); i++ {
+			if file[i] == '/' {
+				topdir := file[:i]
+				if _, ok := dirs[topdir]; !ok {
+					dirs[topdir] = false
+				}
+			}
+		}
+
+		if target != "" {
+			if _, ok := dirs[target]; ok {
+				dirs[file] = false
+			}
+		}
+	}
+	return
+}
+
+// Builds a json from a map of files.
+func AssembleJSON(unmanagedFilesMap interface{}) string {
+	jsonMap := map[string]interface{}{"extracted": false, "files": unmanagedFilesMap}
+	json, _ := json.MarshalIndent(jsonMap, " ", "  ")
+	return string(json)
+}
+
+// Checks if a directory contains any managed directories
+func HasManagedDirs(dir string, rpmDirs map[string]bool) bool {
+	for rpmDir := range rpmDirs {
+		if strings.HasPrefix(rpmDir, dir+"/") {
+			return true
+		}
+	}
+	return false
+}
+
+// Returns unmanaged files in a system.
+func FindUnmanagedFiles(rd func(dir string) ([]os.FileInfo, error), dir string, rpmFiles map[string]string, rpmDirs map[string]bool,
+	unmanagedFiles map[string]string, ignoreList map[string]bool) {
+	files, _ := rd(dir)
+	for _, f := range files {
+		fileName := dir + f.Name()
+		if !utf8.ValidString(fileName) {
+			fmt.Fprintln(os.Stderr, fileName, "contains invalid UTF-8 characters. Skipping.")
+		} else {
+			if _, ok := ignoreList[fileName]; !ok {
+				if f.IsDir() {
+					if _, ok := rpmDirs[fileName]; ok {
+						FindUnmanagedFiles(rd, fileName+"/", rpmFiles, rpmDirs, unmanagedFiles, ignoreList)
+					} else {
+						if !HasManagedDirs(fileName, rpmDirs) {
+							unmanagedFiles[fileName+"/"] = "dir"
+						}
+					}
+				} else {
+					if _, ok := rpmFiles[fileName]; !ok {
+						if f.Mode()&
+							(os.ModeSocket|os.ModeNamedPipe|os.ModeDevice|os.ModeCharDevice) != 0 {
+							// Ignore sockets, named pipes and devices
+						} else if f.Mode()&os.ModeSymlink == os.ModeSymlink {
+							unmanagedFiles[fileName] = "link"
+						} else {
+							unmanagedFiles[fileName] = "file"
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+// Returns managed files and directories in a system.
```

Comment:
comment on exported function GetManagedFiles should be of the form "GetManagedFiles ..."